onnx/utils.py

```diff
@@ -21,3 +24,138 @@ def polish_model(model):  # type: (ModelProto) -> ModelProto
+        names_keep = s_new_names & original_names
+        names_add = s_new_names - original_names
+
+        tensors = []
+        for name in names_keep:
+            tensors.append(tmap[name])
+        for name in names_add:
+            # activation become input or output
+            tensors.append(self.vimap[name])
+
+        # sort the tensors
+        new_tmap = self._buildNameDict(tensors)
+        sorted_tensors = [new_tmap[name] for name in new_names]
+        return sorted_tensors
+
+    def _filterInputs(self, names):  # type: ignore
+        return self._filterTensors(self.graph.input, names)
+
+    def _filterOutputs(self, names):  # type: ignore
+        return self._filterTensors(self.graph.output, names)
+
+    def _dfsSearchReachableNodes(self, node_output_name, graph_input_names, reachable_nodes):  # type: ignore
+        if node_output_name in graph_input_names:
+            return
+        for node in self.graph.node:
+            if node in reachable_nodes:
+                continue
+            if node_output_name not in node.output:
+                continue
+            reachable_nodes.append(node)
+            for name in node.input:
+                self._dfsSearchReachableNodes(name, graph_input_names, reachable_nodes)
+
+    def _filterNodes(self, input_names, output_names):  # type: ignore
+        reachable_nodes = list()  # type: ignore
+        for name in output_names:
+            self._dfsSearchReachableNodes(name, input_names, reachable_nodes)
+        # needs to be topology sorted.
+        nodes = [n for n in self.graph.node if n in reachable_nodes]
+        return nodes
+
+    def _searchReachableTensors(self, nodes):  # type: ignore
+        all_tensors_name = set()
+        for node in nodes:
+            for name in node.input:
+                all_tensors_name.add(name)
+            for name in node.output:
+                all_tensors_name.add(name)
+
+        initializer = [self.wmap[t] for t in self.wmap.keys() if t in all_tensors_name]
+        value_info = [self.vimap[t] for t in self.vimap.keys() if t in all_tensors_name]
+        assert(len(self.graph.sparse_initializer) == 0)
+        assert(len(self.graph.quantization_annotation) == 0)
+        return (initializer, value_info)
+
+    def _make_model(self, nodes, inputs, outputs, initializer, value_info):  # type: ignore
+        name = 'Extracted from {' + self.graph.name + '}'
+        graph = onnx.helper.make_graph(nodes, name, inputs, outputs, initializer=initializer,
+                                      value_info=value_info)
+
+        meta = {
+            'ir_version': self.model.ir_version,
+            'opset_imports': self.model.opset_import,
+            'producer_name': 'onnx.utils.extract',
+        }
+        return onnx.helper.make_model(graph, **meta)
+
+    def extract(self, input_names=None, output_names=None):  # type: ignore
+        graph = self.model.graph
+        if input_names is None:
```

Comment:
if input or output names are not provided then you extract the entire graph. Why not display an error instead?

Reply:
Interestingly, in this case it ends up being a "model optimization" that eliminates nodes that are not used (in computing any of the output values). So, the current behavior seems to have a useful usecase. But it is worth documenting.

Reply:
Remove the _fallback_ like logic - input/output names are taken as they are.

Reply:
For the _model optimization_ like behavior, as we are extracting sub-models - which is removing _unused_ or _unneeded_ nodes and tensors - it seems to be something as expected to me :)